Add size to new files metadata and include it (as well as readme generator) to postUpload script

rija opened the pull request:
# Pull request for issue: #1201

This is a pull request for the following functionalities:

* Implement a console command that update the file size column file table for newly uploaded datasets 
* Refactor a console command that verify that the files urls of a newly updated dataset are valid
* Add to the `postUpload.sh` script, the two commands above, as well as the readme creation command
* Create a configurator project to configure our application without duplicating all the logic in each project

## How to test?

### Locally and manually:

#### Prepration

a. Start the main app
```
$ ./up.sh
```

b. Prepare the files-metadata-console project:

```
$ cd gigadb/app/tools/files-metadata-console
$ ./configure
```

As instructed by the console message, don't forget to provide a correct value for `GITLAB_PRIVATE_TOKEN` in `.env`
Then continue:

```
$ ./configure
$ docker-compose run --rm composer install
```


c. Prepare the readme generator:

```
$ cd gigadb/app/tools/readme-generator
$ cp config-sources/env.example .env
```
Ensure you have provide values for `GITLAB_PRIVATE_TOKEN` and `REPO_NAME` variables.

```
$ docker-compose run --rm tool composer install 
```

#### Execution

Change back to the gigadb-website root directory
```
$ cd ../../../..
```
Then execute postUpload.sh on 100006

```
$ ./gigadb/app/tools/excel-spreadsheet-uploader/postUpload.sh 100006
```

Then check the files generated:
```
$ ls -alrt /tmp/*-100006.txt 
-rw-r--r--  1 owner  wheel   739 Mar 28 17:47 /tmp/updating-md5checksum-100006.txt
-rw-r--r--  1 owner  wheel   504 Mar 28 17:47 /tmp/invalid-urls-100006.txt
-rw-r--r--  1 owner  wheel    15 Mar 28 17:47 /tmp/updating-file-size-100006.txt
-rw-r--r--  1 owner  wheel  2520 Mar 28 17:47 /tmp/readme-100006.txt
```

Opening each file should yield:

```
$ cat /tmp/updating-md5checksum-100006.txt  
No file found at https://ftp.cngb.org/pub/gigadb/pub/10.5524/104001_105000/100006/100006.md5
No file found at https://ftp.cngb.org/pub/gigadb/pub/10.5524/103001_104000/100006/100006.md5
No file found at https://ftp.cngb.org/pub/gigadb/pub/10.5524/102001_103000/100006/100006.md5
No file found at https://ftp.cngb.org/pub/gigadb/pub/10.5524/101001_102000/100006/100006.md5
Processing https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/100006.md5
Saved md5 file attribute with id: 10674
Saved md5 file attribute with id: 10672
Saved md5 file attribute with id: 10673
Saved md5 file attribute with id: 10671
Saved md5 file attribute with id: 10670
Saved md5 file attribute with id: 10669
Saved md5 file attribute with id: 10675
```


```
cat /tmp/invalid-urls-100006.txt  
| URL | Issue |

```

> there is no invalid urls on for DOI, so the file is empty

```
$ cat /tmp/updating-file-size-100006.txt  
nb. changes: 5
```

> That's the number of successful changes

```
cat /tmp/readme-100006.txt   
[DOI] 10.5524/100006

[Title] Genomic data from Adelie penguin (<em>Pygoscelis adeliae</em>). 

[Release Date] 2011-07-06

[...]
```
>That's the readme generated for that dataset.


You can also check on the dataset page (/dataset/100006) that the file size and MD5 checksum are there

#### Variations

* If you want manually to get the tool to pick up on invalid files urls, you can run the following SQL queries against the development database, and then perform the steps from the previous section again:
```
select id from dataset where identifier='100006';
select id,location from file where dataset_id=8;
update file set location='https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006' where id=663;
update file set location='https://example.com/myfile.txt' where id=17677;
update file set location='ftp://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/phylogeny_study_update/Pygoscelis_adeliae.RepeatMasker.out.gz' where id=17679;
update file set location='https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/phylogeny_study_update/' where id=17681;
```

* If you don't pass the DOI to `postUpload.sh`, it should display the usage.


### Locally with automated tests

Run the tests for the file size updater and the url checker: 

```
$ cd gigadb/app/tools/files-metadata-console
$ ./test
```

### Manually on AWS deployment

Ensure you have staging environment provisioned with Terraform and Ansible from this PR's branch.
Push this PR's branch to your remote fork and wait until the following jobs have completed:

* `FilesMetaDataConsoleDeployStaging`
* `ReadmeGeneratorDeployStaging`
* `sd_gigadb`

Then ssh to the bastion server, and run the `postUpload.sh` script like on your local env:

```
$ ./postUpload.sh 100006
```

The output is stored in `/home/centos/uploadLogs`:

```
$ ls -alrt /home/centos/uploadLogs/*-100006.txt
-rw-rw-r--. 1 centos centos   15 Mar 28 16:16 /home/centos/uploadLogs/updating-file-size-100006.txt
-rw-rw-r--. 1 centos centos   16 Mar 28 16:16 /home/centos/uploadLogs/invalid-urls-100006.txt
-rw-rw-r--. 1 centos centos  739 Mar 28 16:16 /home/centos/uploadLogs/updating-md5checksum-100006.txt
-rw-rw-r--. 1 centos centos 2528 Mar 28 16:16 /home/centos/uploadLogs/readme-100006.txt
```
 
You can also check on the dataset page (/dataset/100006) that the file size and MD5 checksum are there.

You can repeat the process on live environment


## How have functionalities been implemented?

A new project was created, called `files-metadata-console` to be the place where you put the code for manipulating files metadata in Postgres using command line interface. It's a Basic Yii2 project, and has two controllers:

* `UpdateController.php`: the console controller for updating file size in database. There is only one Yii2 action (`actionFileSize`).
* `CheckController.php`: the console controller for checking if urls in database are valid. There is only one Yii2 action (`actionValidUrls`)

The domain layer/business logic needed to implement the console command is made of: 

* two components (under the `components/` directory): 
  * `DatasetFilesUpdater.php`: Query the database for list of files, fetch `Content-Length` header from the files url and save the value in the size column for each file in the database
  *  `FilesURLsFetcher.php`: Query the database for list of files, and check whether their url is valid or not and return an array with  the invalid urls alongside the corresponding issue
 * a data layer: both components depends on the Activerecord models at `gigadb/app/models`
 
Both components also rely on a service (located at `gigadb/app/services/URLsService.php`) for batch manipulation of URLs.

## Any issues with implementation?


There is an issue I have discovered last minute: CNGB severs have firewall rules that occasionally make connection to retrieve something on the ftp server (even using the HTTP proxy) fails (and time out) based on:

* IP address of the host making the call
* User agent used by the web client making the call

Rebuilding my environment (and therefore generating new IPs) does tend to fix the issue.
It happened to me on the bastion server, and on my own dev computer (both with IP in Paris)

It might not be a problem on a bastion deployed in Tokyo or HK.

Some changes that could have been done in this PR and weren't due to complexity and/or lack of time:
* the MD5 checksum update function for files sits well in the `files-metadata-console` project
* more test coverage for the configurator
* more test coverage for url checker

## Any changes to automated tests?

* Functional tests in `files-metadata-console` project
* BATS tests for the configurator (see details in `gigadb/app/configurator/README.md`)

## Any changes to documentation?

Added documentation necessary to get all green ticked on https://github.com/gigascience/gigadb-website/community

## Any technical debt repayment?

The files url checker already existed but had issues. It is fixed in this PR.

## Any improvements to CI/CD pipeline?

* configurator project  (see details in `gigadb/app/configurator/README.md`)

rija committed:
feat: Copy yii2 AR model for file table in gigadb/app/models

rija committed:
refactor: move checkUrls out of FilesCommand into DatasetFileService

rija committed:
feat: add yii2 basic console project for managing files metadata

rija committed:
feat: add method to fetch specific header (content length) for URLs

rija committed:
feat: add component and its method to save the size of a dataset's files

rija committed:
fix: remediate failing test on staging

rija committed:
refactor: extract two classes and inject them to domain layer instead

rija committed:
feat: add console controller for updating file size for a dataset

rija committed:
feat: enable docker and allow console to run inside or outside docker

rija committed:
feat: add gitlab test jobs for files-metadata-console

rija committed:
fix: use docker-compose to install composer vendor directory

rija committed:
fix: specify the docker-compose file to use to override global variable

rija committed:
fix: correct invalid type in docker-compose

to fix:
services.composer.environment.DOCKER_RUNNING contains true, which is an invalid type, it should be a string, number, or a null

rija committed:
fix: make tests access database via docker,use docker to run tests on CI

rija committed:
fix: ensure test run use the local docker-compose file

rija committed:
fix: list composer directory on CI for debugging

rija committed:
fix: check pwd on console container for debugging

rija committed:
fix: ensure console and composer service have common working_dir

rija committed:
fix: prevent CI jobs to rebuild the container every time

rija committed:
fix: show images after building the main container

rija committed:
fix: correct container name in gitlab test job for this tool

rija committed:
fix: allow tool to see main docker network and to use cached image

rija committed:
refactor: rename gitlab jobs config for consistency and clarity

rija committed:
fix: pull all images needed by up.sh to avoid rebuilding them

rija committed:
feat: add configuration script and template for DB config

rija committed:
fix: remove ports directive from app service definion

because it's a console app, no need to have ports open

rija committed:
feat: Add Dockerfile, docker-compose for production environments

rija committed:
feat: add new gitlab jobs for building for staging

rija committed:
fix: add missing configure directive on test gitlab job

rija committed:
fix: ensure base images can be loaded on the pipeline jobs

rija committed:
fix: add debug information re location of configurator

rija committed:
fix: wrong directive for specifying DAG and wrong permissions on script

rija committed:
fix: change permission manually in the job

rija committed:
fix: add more debug

rija committed:
fix: more debug

rija committed:
fix: add test file for debug

rija committed:
fix: rename the configurator module

rija committed:
fix: move the sourcing of configurator before failure mode switches

rija committed:
fix: don't always run the staging build job

rija committed:
fix: move export to env inside bash functions

rija committed:
fix: add export -p to debug and use set -a in main script

rija committed:
fix: delete .env from previous jobs

rija committed:
fix: copy .env from project root on CI jobs

rija committed:
fix: add debug

rija committed:
fix: remove local .env to test makeDotEnv in test job

rija committed:
fix: apply previous fix to the staging build job too as it worked

rija committed:
fix: can generate .env on ci

rija committed:
fix: copy .secrets from git repo root

rija committed:
fix: run configure in docker on ci

rija committed:
fix: use GIGADB_ENV to pull current environment from

rija committed:
fix: Also copy .env in from git repo root in addition to .secrets

rija committed:
fix: move accessToken initialisation to if no .secrets block

rija committed:
fix: give correct context to the production docker-compose service

rija committed:
fix: correct typo in container service name

rija committed:
fix: correct the path to the production dockerfile to match context

rija committed:
fix: correct production dockerfile

rija committed:
feat: add gitlab config for staging deploy, live build and deploy

rija committed:
fix: don't always build stating for main app

rija committed:
fix: make staging build and deploy for main app automated but in DAG

rija committed:
fix: on AWS deployment make the DSN accessible to the new configure

rija committed:
fix: remove unecessary build of dev instance of the app

rija committed:
fix: put manual job toggle to treh live build, not in the common block

rija committed:
fix: enable variable export for remote DSN

rija committed:
fix: remove manual toogle from .deploy

rija committed:
fix: correct environment for staging build

rija committed:
fix: persist artefacts in staging deploy for live jobs

rija committed:
fix: save base image explicetely in artifacts

rija committed:
fix: remove dependencies directive with empty array value

rija committed:
refactor: delete unused class

rija committed:
refactor: remove unused code

rija committed:
test: add tests for makeDotEnv()

rija committed:
test: add a basic test for makeDotSecrets()

rija committed:
test: add test runner for the project

rija committed:
fix: use docker-compose inside test runner

rija committed:
fix: set docker-compose files to use in COMPOSE_FILE variable

rija committed:
fix: don't rely on COMPOSE_FILE variable as too wide a scope

rija committed:
refactor: Implement files urls check in files-metadata-console

rija committed:
test: add functional test for check valid urls action

rija committed:
refactor: separate code into controller, component and service

rija committed:
fix: remove redundant code

rija committed:
test: add more tests to cover more scenarios (wrong DOI and happy path)

rija committed:
refactor: remove redundant code

rija committed:
fix: add file size updater and new file url checker to postUpload.sh

rija committed:
feat: add readme generator and guiding messages to postUpload.sh

rija committed:
docs: add README for configurator

rija committed:
feat: show usage if DOI is not passed to postUpload.sh

rija committed:
fix: wrong place for this code

rija committed:
docs: add documentation for community standards

kencho51 commented on gigadb/app/configurator/tests/makeDotSecrets.bats:
```diff
@@ -0,0 +1,20 @@
+    [ "$output" = "Sourcing secrets" ]
+}
+
+# bats test_tags=TODO
+@test "can generate .secret" {
+    skip
+}
```
Hi @rija ,

May be this `TODO` test can be fixed with this patch:
```
@test "can generate .secrets" {
    run makeDotSecrets tests
    [ -s tests/.secrets ] # check if .secrets exists and its size is not zero
}
```

rija replied:
Hi @kencho51,

Good call. I have added the changes.

kencho51 commented on gigadb/app/configurator/src/dotfiles.sh:
```diff
@@ -0,0 +1,84 @@
+  else
+      echo "Neither .env file or default example were present, generating one on the fly"
+      currentEnv=dev
+      username=$(git config --get user.name | cut -d" " -f1 | tr '[:upper:]' '[:lower:]')
```
Hi @rija ,

The username from git config does not necessarily as the same as the gitlab repo name. In my case, I have to change my user name in git config to execute the `./configure`.

Is it necessary to grab the username from git config instead of command line input?

rija replied:
Hi @kencho51,

Your course of action was a bit radical ;-)
You could have simply edited the` .env` file and adjusted the value of REPO_NAME, CI_PROJECT_URL and PROJECT_VARIABLES_URL to match your fork name.

In any case,  I agree that Gitlab fork name cannot be guessed reliably.

However, command line based configuration has to aim to facilitate automation (that was the initial motivation)
so interactive inputs with the user like we have for the Terraform init script is not good at all, and we should not replicate that.

So I have now changed the initial config file creation so that:

by default, `REPO_NAME` behaves like `GITLAB_PRIVATE_TOKEN`: 
It's initially populated with "replace-me" and `makeDotSecrets()` will output an error inviting user to edit .env to correct the value of REPO_NAME (and there is **no** need to change CI_PROJECT_URL and PROJECT_VARIABLES_URL)


You needed to edit the `.env` file anyway to update the private token, so it's not adding new steps to the workflow, and gives you opportunity to review `.env`'s content.

kencho51 replied:
Hi @rija,

Thank you for the update,  updating the `REPO_NAME` in .env together with `GITLAB_PRIVATE_TOKEN` is much simpler now.

kencho51 commented:
Hi @rija,


I wanted to test the scenarios of invalid urls in dev environment, so I updated the urls to:
<img width="1088" alt="image" src="https://user-images.githubusercontent.com/64770635/229721310-16e41be1-a3ba-4d20-8682-5da9d82c0ee6.png">

Then I executed the `postUpload.sh`:
```
% ./gigadb/app/tools/excel-spreadsheet-uploader/postUpload.sh 100006
% ls -alrt /tmp/*-100006.txt              
-rw-r--r--  1 kencho  wheel   751 Apr  4 15:18 /tmp/updating-md5checksum-100006.txt
-rw-r--r--  1 kencho  wheel     0 Apr  4 15:18 /tmp/invalid-urls-100006.txt
-rw-r--r--  1 kencho  wheel     0 Apr  4 15:18 /tmp/updating-file-size-100006.txt
-rw-r--r--  1 kencho  wheel  2571 Apr  4 15:18 /tmp/readme-100006.txt
% cat /tmp/invalid-urls-100006.txt
% cat /tmp/updating-file-size-100006.txt
```
The file `invalid-urls-100006.txt` is empty.

But when I execute the `postUpload.sh 100006` in bastion after the updating the  location of file table in the rds instance with the invalid urls, it works as expected:
<img width="1220" alt="image" src="https://user-images.githubusercontent.com/64770635/229733247-9200bcce-601f-48a9-a178-6b2e47250d34.png">

rija committed:
test: add test that .secrets is generated

rija committed:
fix: Do not guess REPO_NAME, make it behave like GITLAB_PRIVATE_TOKEN

rija commented:
> The file `invalid-urls-100006.txt` is empty.

Hi @kencho51,

I wasn't able to replicate your experience.
From your comment I see the size of your `/tmp/invalid-urls-100006.txt` is zero, which should not happen because even when all urls are correct, it still keep the headers (`| URL | Issue |`)
Can you share the terminal output of your command?

When I run on my local dev:
```
$ ./gigadb/app/tools/excel-spreadsheet-uploader/postUpload.sh 100006
```
I get this output:
```
[...]
* About to check that file urls are valid for 100006
| URL | Issue |
| https://example.com/myfile.txt | Resource cannot be downloaded, not found or forbidden (4xx) |
| ftp://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/phylogeny_study_update/Pygoscelis_adeliae.RepeatMasker.out.gz | Wrong scheme (ftp://) |
| https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006 | URL appears to be a directory listing |
| https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/phylogeny_study_update/ | URL appears to be a directory (/) |

Done with checking that file urls are valid for 100006. Invalid Urls (if any) are save in file: /tmp/invalid-urls-100006.txt

[...]
```

kencho51 commented:
Hi @rija,

> Can you share the terminal output of your command?

Here is the terminal output:
```
% ./gigadb/app/tools/excel-spreadsheet-uploader/postUpload.sh 100006
[...]
* About to check that file urls are valid for 100006
Composer detected issues in your platform:

Your Composer dependencies require a PHP version ">= 8.1.0". You are running 7.3.29-to-be-removed-in-future-macOS.

PHP Fatal error:  Composer detected issues in your platform: Your Composer dependencies require a PHP version ">= 8.1.0". You are running 7.3.29-to-be-removed-in-future-macOS. in /Volumes/kencho/gigadb-website/gigadb/app/tools/files-metadata-console/vendor/composer/platform_check.php on line 24

Done with checking that file urls are valid for 100006. Invalid Urls (if any) are save in file: /tmp/invalid-urls-100006.txt

* About to update files' size for 100006
Composer detected issues in your platform:

Your Composer dependencies require a PHP version ">= 8.1.0". You are running 7.3.29-to-be-removed-in-future-macOS.

PHP Fatal error:  Composer detected issues in your platform: Your Composer dependencies require a PHP version ">= 8.1.0". You are running 7.3.29-to-be-removed-in-future-macOS. in /Volumes/kencho/gigadb-website/gigadb/app/tools/files-metadata-console/vendor/composer/platform_check.php on line 24

Done with updating files' size for 100006. Nb of successful changes saved in file: /tmp/updating-file-size-100006.txt
[...]
% ls -alrt /tmp/*-100006.txt                    
-rw-r--r--  1 kencho  wheel   751 Apr 12 16:06 /tmp/updating-md5checksum-100006.txt
-rw-r--r--  1 kencho  wheel     0 Apr 12 16:06 /tmp/invalid-urls-100006.txt
-rw-r--r--  1 kencho  wheel     0 Apr 12 16:06 /tmp/updating-file-size-100006.txt
-rw-r--r--  1 kencho  wheel  2571 Apr 12 16:06 /tmp/readme-100006.txt

```
The url checking and file size updating are not working in my dev, could it because of this error: `PHP Fatal error:  Composer detected issues in your platform: Your Composer dependencies require a PHP version ">= 8.1.0". You are running 7.3.29-to-be-removed-in-future-macOS. in /Volumes/kencho/gigadb-website/gigadb/app/tools/files-metadata-console/vendor/composer/platform_check.php on line 24` ?

rija committed:
fix: use docker-compose to run commands in postUpload.sh for consistency

rija commented:
Hi @kencho51,

Yes, that's it.
The reason it happens only for size update and url checking, it's because those two commands are the only ones called natively in the `postUpload.sh` script when run on the local dev. environment. 
For consistency, I've now prefixed them in the shell script with the appropriate docker-compose command, so you should not see this problem again.

However, I still recommend you upgrade your native PHP version. `7.3.29-to-be-removed-in-future-macOS` doesn't sound like a good version to have as default system version, and won't work with our newer projects that require PHP 8+, if you like to use the system's PHP (I do).

kencho51 reviewed (approved):
Hi @rija,

**praise:** `postUpload.sh` is working in both staging and dev environment as expected.
**praise:** The test for shell scripts using `bats-core` is implemented.
**praise:** The `TODO` task in `@test "can generate .secrets"` has been fixed.
**praise:** The gitlab `REPO_NAME` can be configured in .env now.
**praise:** The new `makeDotSecrets`and `makeDotEnv`functions make generating the config and secrets in a simpler and more elegant way, which also makes it more maintainable and testable.
**praise:** The url checking and file size updation are now using the docker compose service to execute, independent to the native php installed in dev.
**praise:** All documents for open source community are ready now.

So, happy to approve this PR.

pli888 reviewed (commented):
**praise**: Functionality works on my `dev` environment. I made passing the tests locally easier by VPNing into the BGI network.
**praise**: Functionality works on my `staging` bastion server. It took about 5 minutes for `./postUpload.sh 100006` to successfully execute to finish.
**quibble**: `makeDotEnv.bats` and `makeDotSecrets.bats` contains mix of tabs and spaces indentation.
**issue**: When creating my AWS infrastructure in a clean `ops/infrastructure/envs/staging` directory, I get an error message on executing `terraform plan` that complains about `aws_route_table.database is empty tuple, aws_route_table.private is empty tuple - Call to function "coalescelist" failed: no non-null arguments`.
**suggestion**:I was able to fix above error by uncommenting the `private_subnets` block in `terraform.tf`.

rija committed:
fix: uncomment private subnets block to alleviate Terraform error

rija committed:
fix: add VPC policy to fix Network ACLs error

rija committed:
fix: convert tabs to space

rija commented:
Hi @kencho51, @pli888,

After incorporating @pli888's suggested change, I have added a new set of permissions to the IAM's `Developer` group that relates to VPC life cycle. That fixed my issue.
If you create a new environment from scratch, it should now work.
The additional policy is documented in `docs/awsdocs/policy-vpc.md`.

Also cleaned up the code in the configuration shell scripts.

pli888 commented:
**praise:** With the new changes, I am now able to instantiate an AWS infrastructure for my staging environment without any errors. I confirm the functionality in `postUpload.sh` is still working for me. I approve this PR.

kencho51 commented:
Hi @rija,

**praise:** With the updated VPC permission, I could `terraform apply` my staging instances without any errors.
So, happy to approve this PR.
